server/src/test/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStreamTests.java

```diff
@@ -0,0 +1,95 @@
+
+        assertThat(in.getBytesRead(), equalTo(0));
+
+        final CheckedConsumer<Integer, IOException> readRandomly = (Integer length) -> {
+            switch (between(0, 3)) {
+                case 0 -> {
+                    for (var i = 0; i < length; i++) {
+                        final int bytesBefore = in.getBytesRead();
+                        final int result = in.read();
+                        assertThat((byte) result, equalTo(input[bytesBefore]));
+                        assertThat(in.getBytesRead(), equalTo(bytesBefore + 1));
+                    }
+                }
+                case 1 -> {
+                    final int bytesBefore = in.getBytesRead();
+                    final byte[] b;
+                    if (randomBoolean()) {
+                        b = in.readNBytes(length);
+                    } else {
+                        b = new byte[length];
+                        assertThat(in.read(b), equalTo(length));
+                    }
+                    assertArrayEquals(Arrays.copyOfRange(input, bytesBefore, bytesBefore + length), b);
+                    assertThat(in.getBytesRead(), equalTo(bytesBefore + length));
+                }
+                case 2 -> {
+                    final int bytesBefore = in.getBytesRead();
+                    final byte[] b = new byte[length * between(2, 5)];
+                    if (randomBoolean()) {
+                        assertThat(in.read(b, length / 2, length), equalTo(length));
+                    } else {
+                        assertThat(in.readNBytes(b, length / 2, length), equalTo(length));
+                    }
+                    assertArrayEquals(
+                        Arrays.copyOfRange(input, bytesBefore, bytesBefore + length),
+                        Arrays.copyOfRange(b, length / 2, length / 2 + length)
+                    );
+                    assertThat(in.getBytesRead(), equalTo(bytesBefore + length));
+                }
+                case 3 -> {
+                    final int bytesBefore = in.getBytesRead();
+                    if (randomBoolean()) {
+                        assertThat((int) in.skip(length), equalTo(length));
+                    } else {
+                        in.skipNBytes(length);
+                    }
+                    assertThat(in.getBytesRead(), equalTo(bytesBefore + length));
+                }
+                default -> fail("unexpected");
+            }
+        };
+
+        while (in.getBytesRead() < input.length - 50) {
+            readRandomly.accept(between(1, 30));
+        }
+
+        final int bytesBefore = in.getBytesRead();
+        final byte[] remainingBytes = in.readAllBytes();
+        assertThat(in.getBytesRead(), equalTo(bytesBefore + remainingBytes.length));
+        assertThat(in.getBytesRead(), equalTo(input.length));
+
+        // Read beyond available data has no effect
+        in.read();
+        final byte[] bytes = new byte[between(20, 30)];
```

Comment:
nit Maybe you can set it to zeroes, and assert it's still equal to zeroes at the end of the test.

Reply:
Thanks. Pushed 0436d246f4955181a66352998be82200e64a801e